src/buffer.c

```diff
@@ -240,8 +387,12 @@ size_t
 buffer_pop(struct Buffer *src, void *dst, size_t len) {
     size_t bytes = buffer_peek(src, dst, len);
 
-    if (bytes > 0)
-        advance_read_position(src, bytes);
+    if (bytes > 0) {
+        if (src->type == SOCK_DGRAM)
+            advance_read_position(src, (size_t)ALIGN_16BIT(bytes));
```

Comment:
Maybe we should pull the align bit into `advance_read_position` and `advance_write_position` so we don't need to branch based on type each time.